drivers/gles3/shaders/scene.glsl

```diff
@@ -1244,7 +1256,17 @@ vec4 fog_process(vec3 vertex) {
 	}
 #endif // !DISABLE_LIGHT_DIRECTIONAL
 
-	float fog_amount = 1.0 - exp(min(0.0, -length(vertex) * scene_data.fog_density));
+	float fog_amount = 0.0;
+
+#ifdef USE_DEPTH_FOG
+	float fog_far = scene_data.fog_depth_end > 0.0 ? scene_data.fog_depth_end : scene_data.z_far;
+	float fog_z = smoothstep(scene_data.fog_depth_begin, fog_far, length(vertex));
```

Comment:
Also this will break if `scene_data.fog_depth_begin >= fog_far` unsure if we want to do some safety checks here

As per the description of [`smoothstep`](https://registry.khronos.org/OpenGL-Refpages/gl4/html/smoothstep.xhtml):
> Results are undefined if edge0 ≥ edge1

Reply:
@clayjohn  Any suggestions on that part? When testing, I don't have any problem with the value of fog_depth_end being smaller than fog_depth_begin, in reality it creates this different aspect that can be searched for by devs
As for the value of fog_depth_end to be negative, it just stops fog from working.
Example of fog_depth_end < fog_depth_begin
![image](https://github.com/godotengine/godot/assets/1673249/bb7339de-ff8b-454f-b001-fed200195d90)
Example of fog_depth_end < 0.0
![image](https://github.com/godotengine/godot/assets/1673249/c4ef7f21-8d21-4ec3-982f-f1c342013d59)
(Testing in mobile / forward)

Reply:
The issue is that it might not work universally, the specification says that it's undefined, on some platforms that might mean a crash etc., so I'd suggest adding something ensuring the range on the renderer side, like is done for this:
https://github.com/godotengine/godot/blob/5034478611697358f4e135c69f364ef6d8cd54dc/servers/rendering/renderer_rd/storage_rd/light_storage.cpp#L714

Can you check with depth end and begin being equal?

Reply:
Also is the `far` value of the camera passed at the same time as the other push constant values here? If so id suggest moving the whole check out from the shader for performance, no reason to run a check in the shader that can be run externally when it's run many times in the shader

Reply:
Indeed the first parameter being larger than the second is undefined behaviour in GLSL. In practice, its used widely though and all devices handle it gracefully. I would just leave the code as is. We can patch a fix in if we run into some theoretical new device that is more spec-compliant.